staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/client_test.go

```diff
@@ -0,0 +1,419 @@
+	}{
+		{"AuthByTLS", configTLS},
+		{"AuthByAppRole", configRole},
+	}
+
+	for _, testCase := range configs {
+		client, err := newClientWrapper(testCase.config)
+		if err != nil {
+			t.Fatalf("name: %s, fail to initialize Vault client: %s", testCase.name, err)
+		}
+
+		// Refresh token successfully.
+		err = client.refreshToken(testCase.config)
+		if err != nil {
+			t.Fatalf("name: %s, unexpecte error when refresh token: %s", testCase.name, err)
+		}
+	}
+}
+
+func VaultTestServer(tb testing.TB, handlers map[string]http.Handler) *httptest.Server {
+	mux := http.NewServeMux()
+	if handlers == nil {
+		handlers = DefaultTestHandlers(tb)
+	}
+	for path, handler := range handlers {
+		mux.Handle(path, handler)
+	}
+	server := httptest.NewUnstartedServer(mux)
+
+	cert, err := tls.LoadX509KeyPair(serverCert, serverKey)
+	if err != nil {
+		tb.Fatalf("bad server cert and keys: %s ", err)
+	}
+	certs := []tls.Certificate{cert}
+
+	ca, err := ioutil.ReadFile(cafile)
+	if err != nil {
+		tb.Fatalf("bad ca file: %s ", err)
+	}
+	certPool := x509.NewCertPool()
+	if !certPool.AppendCertsFromPEM(ca) {
+		tb.Fatal("failed to append certificates to pool.")
+	}
+
+	server.TLS = &tls.Config{Certificates: certs, ClientAuth: tls.VerifyClientCertIfGiven, ClientCAs: certPool}
+	server.StartTLS()
+
+	return server
+}
+
+func DefaultTestHandlers(tb testing.TB) map[string]http.Handler {
+	return map[string]http.Handler{
+		"/v1/transit/encrypt/":   &encryptHandler{tb},
+		"/v1/transit/decrypt/":   &decryptHandler{tb},
+		"/v1/auth/cert/login":    &tlsLoginHandler{tb},
+		"/v1/auth/approle/login": &approleLoginHandler{tb},
+	}
+}
+
+type encryptHandler struct {
+	tb testing.TB
+}
+
+// Just prepend "vault:v1:" prefix as encrypted text.
+func (h *encryptHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	token := r.Header.Get("X-Vault-Token")
+	if token == "" {
+		h.tb.Fatal("unauthenticated encrypt request.")
+	}
+
+	msg, err := parseRequest(r)
+	if err != nil {
+		h.tb.Error("error request message for encrypt request: ", err)
+	}
+
+	plain := msg["plaintext"].(string)
+	data := map[string]interface{}{
+		"ciphertext": "vault:v1:" + plain,
+	}
+	buildResponse(w, data)
+}
+
+type decryptHandler struct {
+	tb testing.TB
+}
+
+// Remove the prefix to decrypt the text.
+func (h *decryptHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	token := r.Header.Get("X-Vault-Token")
+	if token == "" {
+		h.tb.Fatal("unauthenticated decrypt request.")
+	}
+
+	msg, err := parseRequest(r)
+	if err != nil {
+		h.tb.Error("error request message for decrypt request: ", err)
+	}
+
+	cipher := msg["ciphertext"].(string)
+	data := map[string]interface{}{
+		"plaintext": strings.TrimPrefix(cipher, "vault:v1:"),
+	}
+	buildResponse(w, data)
+}
+
+type tlsLoginHandler struct {
+	tb testing.TB
+}
+
+// Ensure there is client certificate for tls login
+func (h *tlsLoginHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if len(r.TLS.PeerCertificates) < 1 {
```

Comment:
Can't you just enforce client cert auth on the server?

Reply:
not all requests to the server need client cert in the test case.